syspage.h

```diff
@@ -19,112 +19,130 @@
 #include <hal/hal.h>
 
 
-#define SIZE_APP_NAME 31
-#define SIZE_MAP_NAME 7
+/* syspage_addProg bitflags */
+enum { flagSyspageExec = 0x01 };
 
 
-/* TODO: Make it compatible with Phoenix-RTOS kernel;
- *       take into account map's attributes while data is written to them */
 enum { mAttrRead = 0x01, mAttrWrite = 0x02, mAttrExec = 0x04, mAttrShareable = 0x08,
 	   mAttrCacheable = 0x10, mAttrBufferable = 0x20 };
 
 
-/* syspage_addProg bitflags */
-enum { flagSyspageExec = 0x01 };
+#pragma pack(push, 1)
 
+typedef struct _syspage_seg_t {
+	struct _syspage_seg_t *next, *prev;
 
-/* syspage_validateAddrMap type of validation bitflags */
-enum { flagValidateTop = 0x01, flagValidateMap = 0x02, flagValidateAttr = 0x04 };
+	addr_t addr;
+	size_t fileSz;
+	size_t memSz;
 
+	u32 flags;
+	u32 align;
+} syspage_seg_t;
 
-/* Initialization function */
-extern void syspage_init(void);
 
+typedef struct _syspage_program_t {
+	struct _syspage_program_t *next, *prev;
 
-/* Syspage's location functions */
+	enum { syspage_progBlob = 0, syspage_progSegs } type;
+	union {
+		struct {
+			addr_t start;
+			addr_t end;
+		} blob;
 
-extern void syspage_setAddress(addr_t addr);
+		struct {
+			syspage_seg_t *segs;
+			addr_t entryPoint;
+		} progsegs;
+	};
 
+	char *argv;
 
-extern addr_t syspage_getAddress(void);
+	size_t imapSz;
+	u8 *imaps;
 
+	size_t dmapSz;
+	u8 *dmaps;
+} syspage_program_t;
 
-/* General functions */
 
-extern int syspage_save(void);
+typedef struct _syspage_map_t {
+	struct _syspage_map_t *next, *prev;
 
+	hal_memEntry_t *entries;
 
```

Comment:
I suggest not no prefix type name with `hal_`. This is not inline with convention. I suggest to change name to `mapent_t`